### PR TITLE
Improve hostname calculation logic

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

Improves logic that calculates host names:
- when determining hostname for a freestyle build, use the `Node` instance that runs the build as a fallback
- when determining hostname for a `Node`, use remote shell command execution as a fallback

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

**Review by drodriguezhdez — APPROVED** [review]
LGTM

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 1]
```diff
@@ -750,13 +753,26 @@ public static String getNodeHostname(@Nullable EnvVars envVars, @Nullable Comput
                 if (DatadogUtilities.isValidHostname(computerHostName)) {
                     return computerHostName;
                 }
+
+                Node node = computer.getNode();
+                if (node != null) {
+                    FilePath rootPath = node.getRootPath();
+                    if (rootPath != null) {
+                        ShellCommandCallable hostnameCommand = new ShellCommandCallable(
+                                Collections.emptyMap(), HOSTNAME_CMD_TIMEOUT_MILLIS, "hostname", "-f");
```
Comment (sarah-witt) [inline src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:762]: What do you think about having the same approach we use here, maybe creating a helper method? https://github.com/jenkinsci/datadog-plugin/blob/9e31a06793655958e1cc377bcab97cc9794d5495/src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java#L534

Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:765]: Good point, I agree that it is better to check the OS explicitly. Although reusing the same approach we use for the controller node wouldn't be accurate, as we're accessing a remote worker which may in theory be running a different OS than the controller.
I used a remote callable instead - the same approach that is used in Jenkins internally to determine whether a node is running a Unix-flavoured OS.

Comment (sarah-witt) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:765]: Mhm I see, that makes sense. Do you mind adding some logging so we know when this command fails?

Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:766]: The callable will throw an exception if the command fails. Updated the catch block to print exception along with the computer node name

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 2]
```diff
@@ -772,10 +773,10 @@ public static String getNodeHostname(@Nullable EnvVars envVars, @Nullable Comput
             }
         } catch (InterruptedException e){
             Thread.currentThread().interrupt();
-            logger.fine("Interrupted while trying to extract hostname from StepContext.");
+            logException(logger, Level.FINE, "Interrupted while trying to extract hostname from StepContext.", e);
 
         } catch (Exception e){
-            logger.fine("Unable to extract hostname from StepContext");
+            logException(logger, Level.FINE, "Unable to get hostname for node " + computer.getName(), e);
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:779]: Everything that is done in this `try` block happens inside the `if(computer != null) {` conditional. If computer is null nothing gets executed

[the remark this answers is not in the document]
Comment (sarah-witt) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:779]: Ah that makes sense, good catch!

**Review by sarah-witt — APPROVED** [review]

**Pull request merged**